clippy_lints/src/manual_let_else.rs

```diff
@@ -167,6 +169,50 @@ fn emit_manual_let_else(
     );
 }
 
+/// Returns whether the given let pattern and else body can be turned into a question mark
+///
+/// For this example:
+/// ```ignore
+/// let FooBar { a, b } = if let Some(a) = ex { a } else { return None };
+/// ```
+/// We get as parameters:
+/// ```ignore
+/// pat: Some(a)
+/// else_body: return None
+/// ```
+
+/// And for this example:
+/// ```ignore
+/// let Some(FooBar { a, b }) = ex else { return None };
+/// ```
+/// We get as parameters:
+/// ```ignore
+/// pat: Some(FooBar { a, b })
+/// else_body: return None
+/// ```
+
+/// We output `Some(a)` in the first instance, and `Some(FooBar { a, b })` in the second, because
+/// the question mark operator is applicable here. Callers have to check whether we are in a
+/// constant or not.
+pub fn pat_and_expr_can_be_question_mark<'a, 'hir>(
```

Comment:
+1

Reply:
It feels a little bit specific to `question mark`/`manual_let_else` to me. Should I still move it?

Reply:
I mean, the reason why I put it into a separate function and why it's shared by question_mark and manual_let_else is to make sure that there is definitely no overlap between the two. I don't think it's something that's supposed to be generally usable/useful, compared to say a function that checks if a pattern is refutable or not.

Reply:
As it stands currently it's definitely only useful for those two lints, but this sort of stuff should go in `clippy_utils` regardless. In this case, you give a `Pat` and an `Expr` and get whether it can be represented as a `?`; that, imo, is general enough for it. Even if it's never really used elsewhere it could potentially save someone the effort of reimplementing it, then getting told it was already implemented in another lint

Reply:
Done, I've moved it.